File: coroutine_implement/callback/callback.cpp
#include <fstream>
#include <functional>
#include <future>
#include <iostream>

using CallBack = std::function<void(const std::string &)>;

void async_read(const std::string &filename, CallBack cb) {
  std::async(std::launch::async, [filename, cb]() {
    std::ifstream file(filename);
    if (!file.is_open()) {
      cb("error: unable to open");
      return;
    }

    std::string content((std::istreambuf_iterator<char>(file)),
                        std::istreambuf_iterator<char>());
    file.close();

    cb(content);
  });
}

void async_write(const std::string &filename, const std::string &content,
                 CallBack cb) {
  std::async(std::launch::async, [filename, content, cb]() {
    std::ofstream file(filename);
    if (!file.is_open()) {
      cb("Error: unable to open file");
      return;
    }

    file << content;
    file.close();
    cb("Write successful");
  });
}

int main() {
  async_read("callback.cpp", [](const std::string &content) {
    std::cout << "Read successful\n";
    std::cout << content << std::endl;
  });

  async_write("callback.txt", "Hello, callback",
              [](const std::string &msg) { std::cout << msg << std::endl; });

  std::this_thread::sleep_for(std::chrono::seconds(1));
  return 0;
}

File: seal/coroutine_implement/callback/callback.go
package main

import (
	"fmt"
	"os"
)

func findValue[T any](values ...any) T {
	for _, value := range values {
		if v, ok := value.(T); ok {
			return v
		}
	}
	var zero T
	return zero
}

type Callback func(string, error)

// 同步读取文件
func syncRead(filename string, cb Callback) {
	data, err := os.ReadFile(filename)
	cb(string(data), err)
}

// 异步读取文件
func asyncRead(filename string, cb Callback) {
	cb(string(findValue[[]byte](os.ReadFile(filename))), nil)
}

// 同步写入文件
func syncWrite(filename, content string, cb Callback) {
	err := os.WriteFile(filename, []byte(content), 0644)
	if err != nil {
		cb("", err)
		return
	}
	cb("Write successful", nil)
}

// 异步写入文件
func asyncWrite(filename, content string, cb Callback) {
	cb("Write successful", os.WriteFile(filename, []byte(content), 0644))
}

func main() {
	// 同步读取文件
	syncRead("example.txt", func(content string, err error) {
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		fmt.Println("Read content:", content)
	})

	// 异步读取文件
	asyncRead("example.txt", func(content string, err error) {
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		fmt.Println("Read content:", content)
	})

	// 同步写入文件
	syncWrite("example.txt", "Hello, world!", func(result string, err error) {
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		fmt.Println(result)
	})

	// 异步写入文件
	asyncWrite("example.txt", "Hello, world!", func(result string, err error) {
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		fmt.Println(result)
	})
}
